src/basic/fileio.c

```diff
@@ -706,12 +711,21 @@ int read_full_stream_full(
                         goto finalize;
                 }
 
-                if (n >= READ_FULL_BYTES_MAX) {
+                if (n >= READ_FULL_BYTES_MAX && !FLAGS_SET(flags, READ_FULL_FILE_LARGE_FILE)) {
+                        r = -E2BIG;
+                        goto finalize;
+                }
+
+                if (n >= READ_FULL_LARGE_BYTES_MAX) {
                         r = -E2BIG;
                         goto finalize;
                 }
 
-                n_next = MIN(n * 2, READ_FULL_BYTES_MAX);
+                n_next = n * 2;
+                if (!FLAGS_SET(flags, READ_FULL_FILE_LARGE_FILE))
+                        n_next = MIN(n_next, READ_FULL_BYTES_MAX);
+                else
+                        n_next = MIN(n_next, READ_FULL_LARGE_BYTES_MAX);
```

Comment:
also here, ternary operator might be more straightforward?

Reply:
Yes. Done.